kolibri/core/logger/api.py

```diff
@@ -39,11 +43,374 @@
+                "answer": validated_data["answer"],
+            }
+        return {
+            "type": interaction_types.ANSWER,
+            "answer": validated_data["answer"],
+            "correct": validated_data["correct"],
+        }
+
+    def _update_and_return_mastery_log(self, request, complete, validated_data):
+        if not request.user.is_anonymous():
+            masterylog = MasteryLog.objects.get(
+                user=request.user, mastery_level=validated_data["mastery_level"],
+            )
+            if complete and not masterylog.complete:
+                masterylog.complete = True
+                masterylog.completion_timestamp = validated_data["end_timestamp"]
+                masterylog.save(update_fields=("complete", "completion_timestamp"))
+            return masterylog
+
+    @logger_action
+    def start_attempt(self, request, content_id, context):
+        serializer = StartAttemptSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        session_id = serializer.validated_data["session_id"]
+        item_id = serializer.validated_data["item_id"]
+        start_timestamp = serializer.validated_data["start_timestamp"]
+        end_timestamp = serializer.validated_data["end_timestamp"]
+
+        correct = serializer.validated_data["correct"]
+        complete = serializer.validated_data["complete"]
+
+        answer = serializer.validated_data["answer"]
+        error = serializer.validated_data["error"]
+        hinted = serializer.validated_data["hinted"]
+
+        with transaction.atomic():
+            output = self._do_session_update(request, serializer.validated_data)
+
+            masterylog = self._update_and_return_mastery_log(
+                request, output["complete"], serializer.validated_data
+            )
+
+            interaction = self._generate_interaction(serializer.validated_data)
+
+            attempt = AttemptLog.objects.create(
+                item=item_id,
+                sessionlog_id=session_id,
+                masterylog=masterylog,
+                correct=correct,
+                answer=answer,
+                interaction_history=[interaction],
+                hinted=hinted,
+                error=error,
+                user=request.user,
+                complete=complete,
+                start_timestamp=start_timestamp,
+                completion_timestamp=end_timestamp if complete else None,
+                end_timestamp=end_timestamp,
+            )
+
+            output.update({"attempt_id": attempt.id})
+
+            return output
+
+    @logger_action
+    def update_attempt(self, request, content_id, context):
+        serializer = UpdateAttemptSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        attempt_id = serializer.validated_data["attempt_id"]
+        end_timestamp = serializer.validated_data["end_timestamp"]
+
+        complete = serializer.validated_data["complete"]
```

Comment:
Even though these "actions" are not strictly RESTful from a DB-server perspective, it might be worth exploring making the API a bit more RESTful from a server-client perspective?

For example, we might expose an "attempt" resource, which can take either `POST` to create a new one, or `PUT`/`PATCH` to update it. Under the hood the creation also needs to update some existing rows in certain tables, but abstractly it's still creating a new attempt list, initialized with the first attempt.

In other words, the HTTP API could expose a RESTful logger API with resources that don't strictly map to any particular DB table.

Reply:
Yes, they could definitely be two restful resources here: `session` and `attempt` with a create and update method for each. I was unclear whether that added any particular value to the API or not.

Reply:
When you call this an "action oriented" API, is that basically the same thing as an RPC API, or is there another nuance there?

I think there are some clear advantages of making it RESTful, and also a few disadvantages:

Pros:

* HTTP verbs are well-defined
* Caching and idempotency rules can be easier to reason about
* xAPI is RESTful

Cons:

* Harder to design because we need define new abstract resources
* SCORM is not RESTful